### PR TITLE
[MI-3295] Added API for linking channels with MS Teams

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -442,6 +442,34 @@ func (a *API) getMSTeamsTeamChannels(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if err := body.IsChannelLinkPayloadValid(); err != nil {
+		a.p.API.LogError("Invalid channel link payload.", "Error", err.Error())
+		http.Error(w, "Invalid channel link payload.", http.StatusBadRequest)
+		return
+	}
+
+	if errMsg, statusCode := a.p.LinkChannels(userID, body.MattermostTeamID, body.MattermostChannelID, body.MSTeamsTeamID, body.MSTeamsChannelID); errMsg != "" {
```
Comment (Kshitij-Katiyar) [inline server/api.go:461]: make sure we are logging the error msg in Link Channels if not we can add the logs here

File: server/api_test.go [hunk 2]
```diff
@@ -1153,3 +1153,116 @@ func TestMSTeamsTeamChannels(t *testing.T) {
...
+			},
+			SetupClient: func(c *clientmocks.Client) {
+				c.On("GetChannelInTeam", testutils.GetTeamsTeamID(), testutils.GetTeamsChannelID()).Return(&msteams.Channel{}, nil)
+			},
+			ExpectedResult:     "Channels linked successfully",
+			ExpectedStatusCode: http.StatusOK,
+			Body:               testutils.GetLinkChannelsPayload(testutils.GetTeamID(), testutils.GetChannelID(), testutils.GetTeamsTeamID(), testutils.GetTeamsChannelID()),
+		},
+		{
+			Name: "LinkChannels: error decoding payload",
```
Comment (Kshitij-Katiyar) [inline server/api_test.go:1197]:
```suggestion
			Name: "LinkChannels: error in decoding payload",
```

File: server/api_test.go [hunk 3]
```diff
@@ -1153,3 +1153,116 @@ func TestMSTeamsTeamChannels(t *testing.T) {
...
+			testutils.GetTeamsTeamID()
+			plugin.SetAPI(mockAPI)
+			defer mockAPI.AssertExpectations(t)
+
+			test.SetupPlugin(mockAPI)
+			test.SetupStore(plugin.store.(*storemocks.Store))
+			test.SetupClient(plugin.clientBuilderWithToken("", "", "", "", nil, nil).(*clientmocks.Client))
+
+			w := httptest.NewRecorder()
+			r := httptest.NewRequest(http.MethodPost, "/link-channels", bytes.NewBufferString(test.Body))
```
Comment (Kshitij-Katiyar) [inline server/api_test.go:1252]: constant ?

File: server/getTeamsData.go [hunk 4]
```diff
@@ -132,6 +132,89 @@ func (p *Plugin) GetMSTeamsTeamChannels(teamID, userID string) ([]msteams.Channe
 	return channels, http.StatusOK, nil
 }
 
+func (p *Plugin) LinkChannels(userID, mattermostTeamID, mattermostChannelID, msTeamsTeamID, msTeamsChannelID string) (string, int) {
+	channel, appErr := p.API.GetChannel(mattermostChannelID)
+	if appErr != nil {
+		p.API.LogError("Unable to get the current channel information.", "ChannelID", mattermostChannelID, "Error", appErr.Message)
```
Comment (Kshitij-Katiyar) [inline server/getTeamsData.go:138]:
```suggestion
		p.API.LogError("Unable to get the current channel details.", "ChannelID", mattermostChannelID, "Error", appErr.Message)
```

Comment (Kshitij-Katiyar) [reply server/getTeamsData.go:138]: fix everywhere

File: server/getTeamsData.go [hunk 5]
```diff
@@ -132,6 +132,89 @@ func (p *Plugin) GetMSTeamsTeamChannels(teamID, userID string) ([]msteams.Channe
...
+		return "This team is not enabled for MS Teams sync.", http.StatusForbidden
+	}
+
+	link, err := p.store.GetLinkByChannelID(mattermostChannelID)
+	if err != nil {
+		p.API.LogError("Error occurred while getting channel link with Mattermost channelID.", "ChannelID", mattermostChannelID, "Error", err.Error())
+		return "Error occurred while getting channel link with Mattermost channelID.", http.StatusInternalServerError
+	}
+	if link != nil {
+		return "A link for this channel already exists. Please unlink the channel before you link again with another channel.", http.StatusBadRequest
```
Comment (Kshitij-Katiyar) [inline server/getTeamsData.go:163]: Please unlink the channel first before linking again

Comment (Kshitij-Katiyar) [reply server/getTeamsData.go:163]: fix this everywhere

**Review by ayusht2810 — APPROVED** [review]

File: server/api.go (hunk 1 above)
Comment (raghavaggarwal2308) [reply server/api.go:465]: yes we are!

File: server/api_test.go (hunk 3 above)
Comment (raghavaggarwal2308) [reply server/api_test.go:1254]: discussed

File: server/getTeamsData.go (hunk 5 above)
Comment (raghavaggarwal2308) [reply server/getTeamsData.go]: I don't think what we have is fine

File: server/api.go [hunk 6]
```diff
@@ -446,6 +446,34 @@ func (a *API) getMSTeamsTeamChannels(w http.ResponseWriter, r *http.Request) {
 	a.writeJSONArray(w, http.StatusOK, paginatedChannels)
 }
 
+func (a *API) linkChannels(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get(HeaderMattermostUserID)
+
+	var body *storemodels.ChannelLink
+	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
+		a.p.API.LogError("Error occurred while decoding link channels payload.", "Error", err.Error())
+		http.Error(w, "Error occurred while decoding link channels payload.", http.StatusInternalServerError)
```
Comment (manojmalik20) [inline server/api.go:455]:
```suggestion
		a.p.API.LogError("Error occurred while unmarshaling link channels payload.", "Error", err.Error())
		http.Error(w, "Error occurred while unmarshaling link channels payload.", http.StatusBadRequest)
```

File: server/api.go [hunk 7]
```diff
@@ -446,6 +446,34 @@ func (a *API) getMSTeamsTeamChannels(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, "Invalid channel link payload.", http.StatusBadRequest)
+		return
+	}
+
+	if errMsg, statusCode := a.p.LinkChannels(userID, body.MattermostTeamID, body.MattermostChannelID, body.MSTeamsTeamID, body.MSTeamsChannelID); errMsg != "" {
+		http.Error(w, errMsg, statusCode)
+		return
+	}
+
+	if _, err := w.Write([]byte("Channels linked successfully")); err != nil {
```
Comment (manojmalik20) [inline server/api.go:470]: Should we return `201` status code here, as a new link is created? :thinking:

File: server/api_test.go [hunk 8]
```diff
@@ -1155,3 +1155,116 @@ func TestMSTeamsTeamChannels(t *testing.T) {
...
+				}, nil).Times(1)
+				api.On("GetConfig").Return(&model.Config{
+					ServiceSettings: model.ServiceSettings{
+						SiteURL: model.NewString("/"),
+					},
+				}, nil).Times(2)
+				api.On("HasPermissionToChannel", testutils.GetUserID(), testutils.GetChannelID(), model.PermissionManageChannelRoles).Return(true).Times(1)
+			},
+			SetupStore: func(store *storemocks.Store) {
+				store.On("GetTokenForMattermostUser", testutils.GetUserID()).Return(&oauth2.Token{}, nil).Times(2)
```
Comment (manojmalik20) [inline server/api_test.go:1183]: This should be called only once.

File: server/api_test.go [hunk 9]
```diff
@@ -1155,3 +1155,116 @@ func TestMSTeamsTeamChannels(t *testing.T) {
...
+				store.On("CheckEnabledTeamByTeamID", testutils.GetTeamID()).Return(true).Times(1)
+				store.On("GetLinkByChannelID", testutils.GetChannelID()).Return(nil, nil).Times(1)
+				store.On("GetLinkByMSTeamsChannelID", testutils.GetTeamsTeamID(), testutils.GetTeamsChannelID()).Return(nil, nil).Times(1)
+				store.On("GetTokenForMattermostUser", testutils.GetUserID()).Return(&oauth2.Token{}, nil).Times(1)
```
Comment (manojmalik20) [inline server/api_test.go:1187]: This is already present above.

File: server/api_test.go [hunk 10]
```diff
@@ -1155,3 +1155,116 @@ func TestMSTeamsTeamChannels(t *testing.T) {
...
+			},
+			SetupClient: func(c *clientmocks.Client) {
+				c.On("GetChannelInTeam", testutils.GetTeamsTeamID(), testutils.GetTeamsChannelID()).Return(&msteams.Channel{}, nil)
+			},
+			ExpectedResult:     "Channels linked successfully",
+			ExpectedStatusCode: http.StatusOK,
+			Body:               testutils.GetLinkChannelsPayload(testutils.GetTeamID(), testutils.GetChannelID(), testutils.GetTeamsTeamID(), testutils.GetTeamsChannelID()),
+		},
+		{
+			Name: "LinkChannels: error in decoding payload",
```
Comment (manojmalik20) [inline server/api_test.go:1199]:
```suggestion
			Name: "LinkChannels: error in unmarshaling payload",
```

File: server/api_test.go [hunk 11]
```diff
@@ -1155,3 +1155,116 @@ func TestMSTeamsTeamChannels(t *testing.T) {
...
+			},
+			SetupClient:        func(c *clientmocks.Client) {},
+			ExpectedResult:     "Invalid channel link payload.\n",
+			ExpectedStatusCode: http.StatusBadRequest,
+			Body:               testutils.GetLinkChannelsPayload("", testutils.GetChannelID(), testutils.GetTeamsTeamID(), testutils.GetTeamsChannelID()),
+		},
+		{
+			Name: "LinkChannels: error occurred while linking channels",
+			SetupPlugin: func(api *plugintest.API) {
+				api.On("GetChannel", testutils.GetChannelID()).Return(nil, &model.AppError{Message: "error occurred while linking channels"}).Times(1)
```
Comment (manojmalik20) [inline server/api_test.go:1229]: The error message is not matching with the function.

File: server/api_test.go [hunk 12]
```diff
@@ -1155,3 +1155,116 @@ func TestMSTeamsTeamChannels(t *testing.T) {
...
+			assert := assert.New(t)
+			plugin := newTestPlugin(t)
+			mockAPI := &plugintest.API{}
+			testutils.GetTeamsTeamID()
+			plugin.SetAPI(mockAPI)
+			defer mockAPI.AssertExpectations(t)
+
+			test.SetupPlugin(mockAPI)
+			test.SetupStore(plugin.store.(*storemocks.Store))
+			test.SetupClient(plugin.clientBuilderWithToken("", "", "", "", nil, nil).(*clientmocks.Client))
```
Comment (manojmalik20) [inline server/api_test.go:1251]: Same as comment on last PR.

File: server/command_test.go [hunk 13]
```diff
@@ -685,8 +685,9 @@ func TestExecuteLinkCommand(t *testing.T) {
 				api.On("GetChannel", "").Return(nil, testutils.GetInternalServerAppError("Error while getting the current channel.")).Times(1)
 				api.On("SendEphemeralPost", "", &model.Post{
 					UserId:  "bot-user-id",
-					Message: "Unable to get the current channel information.",
+					Message: "Unable to get the current channel details.",
 				}).Return(testutils.GetPost("", "")).Times(1)
+				api.On("LogError", "Unable to get the current channel details.", "ChannelID", "", "Error", "").Times(1)
```
Comment (manojmalik20) [inline server/command_test.go:690]: Why are channelID and error both empty?

File: server/get_teams_data.go [hunk 14]
```diff
@@ -126,6 +128,89 @@ func (p *Plugin) GetMSTeamsTeamChannels(teamID, userID string) ([]msteams.Channe
...
+func (p *Plugin) LinkChannels(userID, mattermostTeamID, mattermostChannelID, msTeamsTeamID, msTeamsChannelID string) (responseMsg string, statusCode int) {
+	channel, appErr := p.API.GetChannel(mattermostChannelID)
+	if appErr != nil {
+		p.API.LogError("Unable to get the current channel details.", "ChannelID", mattermostChannelID, "Error", appErr.Message)
+		return "Unable to get the current channel details.", http.StatusInternalServerError
+	}
+
+	if channel.Type == model.ChannelTypeDirect || channel.Type == model.ChannelTypeGroup {
+		p.API.LogError("Linking/unlinking a direct or group message is not allowed.", "ChannelType", channel.Type)
+		return "Linking/unlinking a direct or group message is not allowed", http.StatusForbidden
```
Comment (manojmalik20) [inline server/get_teams_data.go:140]:
```suggestion
		return "Linking/unlinking a direct or group message is not allowed", http.StatusBadRequest
```

File: server/get_teams_data.go [hunk 15]
```diff
@@ -126,6 +128,89 @@ func (p *Plugin) GetMSTeamsTeamChannels(teamID, userID string) ([]msteams.Channe
...
+	}
+
+	if !p.API.HasPermissionToChannel(userID, mattermostChannelID, model.PermissionManageChannelRoles) {
+		p.API.LogError("Unable to link the channel. You have to be a channel admin to link it.", "ChannelID", mattermostChannelID)
+		return "Unable to link the channel. You have to be a channel admin to link it.", http.StatusForbidden
+	}
+
+	if !p.store.CheckEnabledTeamByTeamID(mattermostTeamID) {
+		p.API.LogError("This team is not enabled for MS Teams sync.", "TeamID", mattermostTeamID)
+		return "This team is not enabled for MS Teams sync.", http.StatusForbidden
```
Comment (manojmalik20) [inline server/get_teams_data.go:150]:
```suggestion
		return "This team is not enabled for MS Teams sync.", http.StatusBadRequest
```

File: server/get_teams_data.go [hunk 16]
```diff
@@ -126,6 +128,89 @@ func (p *Plugin) GetMSTeamsTeamChannels(teamID, userID string) ([]msteams.Channe
...
+		return "A link for this channel already exists. Please unlink the channel before you link again with another channel.", http.StatusBadRequest
+	}
+
+	link, err = p.store.GetLinkByMSTeamsChannelID(msTeamsTeamID, msTeamsChannelID)
+	if err != nil && err.Error() != ErrorNoRowsInResult {
+		p.API.LogError("Error occurred while getting the channel link with MS Teams channel ID", "Error", err.Error())
+		return "Error occurred while getting the channel link with MS Teams channel ID", http.StatusInternalServerError
+	}
+	if link != nil {
+		return "A link for this channel already exists. Please unlink the channel before you link again with another channel.", http.StatusBadRequest
```
Comment (manojmalik20) [inline server/get_teams_data.go:168]: The error message should be different in this case.
Something like, "The Teams channel that you're trying to link is already linked to another Mattermost channel. Please unlink that channel and try again."

File: server/get_teams_data.go [hunk 17]
```diff
@@ -126,6 +128,89 @@ func (p *Plugin) GetMSTeamsTeamChannels(teamID, userID string) ([]msteams.Channe
...
+	}
+
+	client, err := p.GetClientForUser(userID)
+	if err != nil {
+		p.API.LogError("Error occurred while getting client for the user.", "Error", err.Error())
+		return "Unable to link the channel, looks like your account is not connected to MS Teams", http.StatusInternalServerError
```
Comment (manojmalik20) [inline server/get_teams_data.go:174]:
```suggestion
		return "Unable to link the channel, looks like your account is not connected to MS Teams", http.StatusUnauthorized
```
either `400` or `401`

File: server/store/storemodels/storemodels.go [hunk 18]
```diff
@@ -43,3 +46,19 @@ type ChannelSubscription struct {
...
 	Secret         string
 }
+
+func (c *ChannelLink) IsChannelLinkPayloadValid() error {
+	if c.MattermostChannelID == "" {
+		return errors.New("mattermost channel ID is required")
+	}
+	if c.MattermostTeamID == "" {
+		return errors.New("mattermost team ID is required")
+	}
```
Comment (manojmalik20) [inline server/store/storemodels/storemodels.go:56]: Reverse the order.

File: server/store/storemodels/storemodels.go [hunk 19]
```diff
@@ -43,3 +46,19 @@ type ChannelSubscription struct {
...
+	if c.MSTeamsChannelID == "" {
+		return errors.New("ms teams channel ID is required")
+	}
+	if c.MSTeamsTeamID == "" {
+		return errors.New("ms teams team ID is required")
+	}
```
Comment (manojmalik20) [inline server/store/storemodels/storemodels.go:62]: Same.

File: server/api.go [hunk 20]
```diff
@@ -454,8 +457,15 @@ func (a *API) linkChannels(w http.ResponseWriter, r *http.Request) {
 
 	var body *storemodels.ChannelLink
 	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
-		a.p.API.LogError("Error occurred while decoding link channels payload.", "Error", err.Error())
-		http.Error(w, "Error occurred while decoding link channels payload.", http.StatusInternalServerError)
+		a.p.API.LogError("Error occurred while unmarshaling link channels payload.", "Error", err.Error())
+		http.Error(w, "Error occurred while unmarshaling link channels payload.", http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
```
Comment (manojmalik20) [inline server/api.go:464]: I don't think there's a need for this.

File: server/api.go [hunk 21]
```diff
@@ -454,8 +457,15 @@ func (a *API) linkChannels(w http.ResponseWriter, r *http.Request) {
...
+
+	if body == nil {
```
Comment (manojmalik20) [inline server/api.go:466]: This should be inside the `IsValid` function. Discuss with Kshitij on this.

File: server/get_teams_data.go [hunk 22]
```diff
@@ -97,17 +97,14 @@ func (p *Plugin) GetMSTeamsChannelDetailsForAllTeams(msTeamsTeamIDsVsChannelsQue
 	return errorsFound
 }
 
-func (p *Plugin) GetMSTeamsTeamList(userID string, r *http.Request) ([]*msteams.Team, int, error) {
-	var client msteams.Client
+func (p *Plugin) GetMSTeamsTeamList(userID string, client msteams.Client) ([]*msteams.Team, int, error) {
 	var err error
-	if r.Context().Value(ContextClientKey) == nil {
+	if client == nil {
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/get_teams_data.go:102]: resolved in PR #20

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**